src/app/modules/courses-list/components/search/search.component.ts

```diff
@@ -8,12 +10,23 @@ import {CoursesService} from '../../../../core/services/courses/courses.service'
   changeDetection: ChangeDetectionStrategy.OnPush
 })
 
-export class SearchComponent {
-  searchText: string;
+export class SearchComponent implements OnInit {
+  private searchChanged$ = new Subject<string>();
 
   constructor(private coursesService: CoursesService) { }
 
-  filterCourses(): void {
-    this.coursesService.filterCourses(this.searchText);
+  ngOnInit() {
+    this.searchChanged$
+      .pipe(
+        debounceTime(500),
+        filter(value => value.length >= 3 || value.length === 0)
+      )
+      .subscribe((value) => {
```

Comment:
Also I propose to add `distinctuntilchanged` method for checking if previous query is different from current